File: Answer/3043.最长公共前缀的长度.py
#
# @lc app=leetcode.cn id=3043 lang=python3
#
# [3043] 最长公共前缀的长度
#

# @lc code=start
from typing import List


class Solution:
    def longestCommonPrefix(self, arr1: List[int], arr2: List[int]) -> int:
        prefix_set = set()

        for num in arr1:
            n = num
            while n > 0:
                prefix_set.add(n)
                n //= 10

        max_len = 0
        for num in arr2:
            length = 0
            temp = num
            while temp > 0:
                length += 1
                temp //= 10

            n = num
            while n > 0:
                if n in prefix_set:
                    if length > max_len:
                        max_len = length
                    break
                length -= 1
                n //= 10

        return max_len
# @lc code=end

File: Answer/3043.最长公共前缀的长度.ts
/*
 * @lc app=leetcode.cn id=3043 lang=typescript
 *
 * [3043] 最长公共前缀的长度
 */

// @lc code=start
function longestCommonPrefix(arr1: number[], arr2: number[]): number {
  const prefixSet = new Set<number>();

  for (const num of arr1) {
    for (let n = num; n > 0; n = Math.floor(n / 10)) {
      prefixSet.add(n);
    }
  }

  let maxLen = 0;
  for (const num of arr2) {
    let length = 0;
    for (let temp = num; temp > 0; temp = Math.floor(temp / 10)) {
      length++;
    }

    for (let n = num; n > 0; n = Math.floor(n / 10)) {
      if (prefixSet.has(n)) {
        if (length > maxLen) {
          maxLen = length;
        }
        break;
      }
      length--;
    }
  }

  return maxLen;
}
// @lc code=end
